radicle-seed/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+#[macro_use]
+extern crate async_trait;
+
+use std::{collections::HashSet, net::SocketAddr, path::PathBuf, vec};
+use thiserror::Error;
+
+use radicle_keystore::sign::ed25519;
+
+use librad::{
+    git,
+    keys,
+    meta::project,
+    net::{
+        discovery,
+        gossip,
+        gossip::types::PeerInfo,
+        peer::{self, PeerApi, PeerConfig},
+        protocol::ProtocolEvent,
+    },
+    paths,
+    peer::PeerId,
+    uri::RadUrn,
+};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Storage(#[from] git::storage::Error),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    #[error(transparent)]
+    Bootstrap(#[from] peer::BootstrapError),
+
+    #[error(transparent)]
+    Accept(#[from] peer::AcceptError),
+}
+
+#[derive(Clone)]
+pub struct Signer {
+    key: keys::SecretKey,
+}
+
+impl Signer {
+    fn new() -> Self {
+        let key = keys::SecretKey::new();
+        Self { key }
+    }
+}
+
+#[async_trait]
+impl ed25519::Signer for Signer {
+    type Error = std::convert::Infallible;
+
+    fn public_key(&self) -> ed25519::PublicKey {
+        self.key.public_key()
+    }
+
+    async fn sign(&self, data: &[u8]) -> Result<ed25519::Signature, Self::Error> {
+        <keys::SecretKey as ed25519::Signer>::sign(&self.key, data).await
+    }
+}
+
+impl keys::AsPKCS8 for Signer {
+    fn as_pkcs8(&self) -> Vec<u8> {
+        self.key.as_pkcs8()
+    }
+}
+
+#[derive(Debug)]
+pub enum Mode {
+    /// Track everything we see, no matter where it comes from.
+    TrackEverything,
+    /// Track everything from these peers, and nothing else.
+    TrackPeers(HashSet<PeerId>),
+    /// Track the specified URNs.
+    TrackUrns(HashSet<RadUrn>),
+}
+
+impl Mode {
+    fn is_trackable(&self, peer: &PeerId, urn: &RadUrn) -> bool {
```

Comment:
Would it make more sense to have `Either<&PeerId, &RadUrn>`? Like, why would we want to have to pass a `urn` if just want to check if a `peer` is trackable?

Reply:
Because in the end, it's the URN that you are tracking, not the peer. So you always want to be passing in the URN

Reply:
Ah, I think I'm following. And you always have a `PeerId` because you need to ask for this from some peer? Or should _that_ be optional?

Reply:
You always have it too in practice. You could *theoretically* have a bare URN, but can't think of a useful context where that would occur.